src/pybind/mgr/volumes/fs/stats_util.py

```diff
@@ -54,3 +75,246 @@ def get_stats(src_path, dst_path, fs_handle):
         'amount cloned': get_size_ratio_str(size_c, size_t),
         'files cloned': get_num_ratio_str(rentries_c, rentries_t),
     }
+
+
+class CloneInfo:
+
+    def __init__(self, volname):
+        self.volname = volname
+
+        self.src_group_name = None
+        self.src_subvol_name = None
+        self.src_path = None
+
+        self.dst_group_name = None
+        self.dst_subvol_name = None
+        self.dst_path = None
+
+
+class CloneProgressReporter:
+
+    def __init__(self, volclient, vol_spec):
+        self.vol_spec = vol_spec
+
+        # instance of VolumeClient is needed here so that call to
+        # LibCephFS.getxattr() can be made.
+        self.volclient = volclient
+
+        # need to figure out how many progress bars should be printed. print 1
+        # progress bar if number of ongoing clones is less than this value,
+        # else print 2.
+        self.max_concurrent_clones = self.volclient.mgr.max_concurrent_clones
+
+        # Creating an RTimer instance in advance so that we can check if clone
+        # reporting has already been initiated by calling RTimer.is_alive().
+        self.update_task = RTimer(1, self._update_progress_bars)
+
+    def initiate_reporting(self):
+        log.info('initiating progress reporting for clones...')
+        if self.update_task.is_alive():
+            return
+
+        # progress event ID for ongoing clone jobs
+        self.on_pev_id: Optional[str] = str(uuid4())
+        # progress event ID for ongoing+pending clone jobs
+        self.onpen_pev_id: Optional[str] = str(uuid4())
+
+        self.update_task = RTimer(1, self._update_progress_bars)
```

Comment:
Yes, I replied to it. I am keeping it `_update_progress_bars` because it's an internal method for this class, not meant to be used by code outside this method.

Reply:
See: https://github.com/ceph/ceph/pull/54620#discussion_r1645644525

Reply:
Than you for the link.

Reply:
Sorry, I don't get it. I've replied to both comments from you there. Am I missing something?

![image](https://github.com/ceph/ceph/assets/10697359/67f2005f-d544-4f05-b889-91cab303cd90)

Reply:
I'll prefer keeping keeping underscore preceding methods names so that users of this class in future will intuitively know which methods are meant for calling through object and which are not. Let me know if you like to overrule that, I'll remove preceding underscores.

Reply:
Well, in that case, almost all routines expect 1 should start with an "_". IMO, the point of using "_" prefixed function names is to make the caller think a little more as it why calling it from a class object is really required. This is pretty relaxed in python and is why I have a bit of a shrug when it comes to things like this.

Anyway, I suggest audit it not more than once and if possible try to make the usage if "_ prefixes routines minimal. I stand to be corrected though :)

Reply:
> almost all routines expect 1 should start with an _.

I understand but that's only a co-incidence. I've tried to keep follow the Python convention so that future users can tell what methods are recommended to call and what are not.

Following are name of all the methods right now -

```
__init__
initiate_reporting
_get_clone_dst_info
_get_clone_src_info
_get_info_for_all_clones
_update_progress_bar_event
_update_progress_bars
_finish_progress_events
finish
```

Calling `initiate_reporting()` and `finish()` makes sense but I don't see why woud user want to call any other methods because all of them are helper to `_update_progress_bars()`.